src/relax/transform/gradient.cc

```diff
@@ -0,0 +1,405 @@
+      return ReplaceExprByVar(src2);
+    } else if (zeros_tracker_.count(src2) != 0) {
+      return ReplaceExprByVar(src1);
+    }
+
+    if (const auto* node1 = src1.as<TupleNode>()) {
+      if (const auto* node2 = src2.as<TupleNode>()) {
+        ICHECK(node1->fields.size() == node2->fields.size()) << "size of tuple not match";
+        Array<Expr> result;
+        for (size_t i = 0; i < node1->fields.size(); ++i) {
+          result.push_back(DoAdd(node1->fields[i], node2->fields[i]));
+        }
+        return Tuple(result);
+      } else {
+        LOG(FATAL) << "Type not match: src1 and src2 should be both tuple";
+        return Expr();
+      }
+    } else {
+      return Call(add_op_, {src1, ReplaceExprByVar(src2)});
+    }
+  }
+
+  Tuple DoAddInTuple(const Tuple& origin, int index, const Expr& increment) {
+    Array<Expr> ret;
+    for (size_t i = 0; i < origin->fields.size(); ++i) {
+      if ((int)i == index) {
+        ret.push_back(DoAdd(origin->fields[i], increment));
+      } else {
+        ret.push_back(origin->fields[i]);
+      }
+    }
+    return Tuple(ret);
+  }
+
+  void BindAndEmit(Var v, Expr e) {
+    if (adjoint_expr_to_var_.count(e)) {
+      e = adjoint_expr_to_var_[e];
+    } else {
+      adjoint_expr_to_var_.Set(e, v);
+      // e->checked_type_ = v->checked_type();
+      // e->shape_ = v->shape();
+    }
+    if (v->IsInstance<DataflowVarNode>()) {
+      builder_->Emit(VarBinding(v, e));
+    } else {
+      builder_->EmitOutput(VarBinding(v, e));
+    }
+  }
+
+  void CheckTarget(const Expr& e) {
+    ICHECK(!e->IsInstance<DataflowVarNode>()) << "not an output node";
+    ICHECK(e->checked_type().as<DynTensorTypeNode>()) << "target must be a DynTensorType";
+    ICHECK(e->shape().as<ShapeExprNode>()) << "error when getting target shape";
+    const auto* shape_node = e->shape().as<ShapeExprNode>();
+    ICHECK(shape_node->values.size() == 0) << "target must be a scalar";
+  }
+
+  void InitGrad(const Var& adjoint_var, const Var& var) {
+    ObjectPtr<InitAttrs> attrs = make_object<InitAttrs>();
+    auto type = Downcast<DynTensorType>(var->checked_type());
+    attrs->dtype = type->dtype;
+
+    const Expr& init = Call(ones_op_, {var->shape()}, Attrs(attrs));
+    adjoint_expr_map_.Set(var, init);
+  }
+
+  // the differentiation target
+  Var target_var_;
+  // the arguments to differentiate
+  Array<Var> require_grads_;
+
+  // var to its adjoints var
+  Map<Var, Var> adjoint_var_map_;
+  // var to its adjoint expr
+  Map<Var, Expr> adjoint_expr_map_;
+  // trace binding
+  Map<Expr, Var> adjoint_expr_to_var_;
+  // track zeros introduced
+  std::unordered_set<Expr, ObjectPtrHash, ObjectPtrEqual> zeros_tracker_;
+
+  // gop map
+  const OpAttrMap<FPrimalGradient> gradient_op_map_ =
+      Op::GetAttrMap<FPrimalGradient>("FPrimalGradient");
+
+  // constant
+  const Op& ones_op_ = Op::Get("relax.ones");
+  const Op& add_op_ = Op::Get("relax.add");
+  const Op& zeros_op_ = Op::Get("relax.zeros");
```

Comment:
They can also be static right?

Reply:
Yep, but if we change it to be static, the initialization should be out of class due to C++ restriction. Like this:

```
class GradientMutator{
//....
  static const Op& ones_op_;
};
const Op& GradientMutator::ones_op_ = Op::Get("relax.ones");
```

So which solution is better?

Reply:
And I am not certain for that, the registration of "relax.ones" happens in some C++ compilation unit statically, and the use of it `Op::Get("relax.ones")` happens in another compilation unit, also statically. It seems that to correctly find the operator, the first should take place before the second. Could that order be guaranteed?

In my trial it works well, though.

Reply:
I see. I noticed that `ones_op_` is only used in `InitGrad`, `add_op_` is only used in `DoAdd` and `zeros_op_` are only used in two places. Given they are not widely used throughout the class, I think it’s fine to put them locally in each method and mark them as constant static.